plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsStepContextTest.java

```diff
@@ -0,0 +1,203 @@
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.LoggerRule;
+import org.jvnet.hudson.test.TestExtension;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.util.Set;
+import java.util.logging.Level;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.Matchers.not;
+
+public class CpsStepContextTest {
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Rule
+    public LoggerRule logger = new LoggerRule();
+
+    @Issue("JENKINS-75067")
+    @Test
+    public void failingStepListenerNotLeakClosures() throws Exception {
+        // Even before the fix there's only one warning logged.  Asserting zero records is probably over-stepping,
+        // but asserting just one record with our target message risks a false negative (some other unrelated message
+        // being first, and our being later).
+        logger.record(CpsThreadGroup.class, Level.WARNING).capture(10);
+        WorkflowJob job = r.createProject(WorkflowJob.class, "p");
+        job.setDefinition(new CpsFlowDefinition("node {}\n", true));
+
+        WorkflowRun build = r.buildAndAssertStatus(Result.FAILURE, job);
+        r.assertLogContains("oops", build);
+        assertThat(ClosureCounter.get().closureCount, equalTo(0));
+        assertThat(logger.getMessages(), not(hasItem(containsString("Stale closure"))));
+    }
+
+    @TestExtension("failingStepListenerNotLeakClosures")
+    public static class FailingStepListener implements StepListener {
+
+        @Override
+        public void notifyOfNewStep(@NonNull Step s, @NonNull StepContext context) {
+            context.onFailure(new AbortException("oops"));
+        }
+    }
+
+    @Issue("JENKINS-75067")
+    @Test
+    public void executionStartExceptionNotLeakClosures() throws Exception {
+        logger.record(CpsThreadGroup.class, Level.WARNING).capture(10);
+        WorkflowJob job = r.createProject(WorkflowJob.class, "p");
+        job.setDefinition(new CpsFlowDefinition("badBlock {}\n", true));
+
+        WorkflowRun build = r.buildAndAssertStatus(Result.FAILURE, job);
+        r.assertLogContains("oops", build);
+        assertThat(ClosureCounter.get().closureCount, equalTo(0));
+        assertThat(logger.getMessages(), not(hasItem(containsString("Stale closure"))));
+    }
+
+    @Issue("JENKINS-75067")
+    @Test
+    public void executionWithBodyRunningSyncNotLeakClosures() throws Exception {
+        logger.record(CpsThreadGroup.class, Level.WARNING).capture(10);
+        WorkflowJob job = r.createProject(WorkflowJob.class, "p");
+        job.setDefinition(new CpsFlowDefinition("echo passthrough {}\n", true));
```

Comment:
Not clear to me from reading how this even parses. Is it

```groovy
echo(passthrough {})
```

? If so, might be clearer to write e.g.
```suggestion
        job.setDefinition(new CpsFlowDefinition("def r = passthrough {}; echo r", true));
```

Reply:
Yes, that is how it is interpreted. I thought about recommending the same change for clarity.

Reply:
Good call.  I was actually going to remove the `echo` (and the return of an actual value from the step), but originally wanted the confidence of knowing the step actually executed (via build log assertion).  I simply forgot to remove them when I saw it.  Since you didn't question the value of those, I'll keep them and use the more explicit change.